integration-testing.md

```diff
@@ -0,0 +1,606 @@
+the issue, change to the `dev/tests/integration` directory, adjust any relative paths accordingly, and run the tests
+again.
+
+<a href="#unable-to-connect-to-mysql"></a>
+### Unable to Connect to MySQL
+
+The PHP interpreter must be able to connect to the test database. By default, this means the tests have to run on the
+same host as the MySQL server. This problem most commonly occurs during development with Vagrant or Docker, where the
+Magento database is running on a virtual machine. If the tests then are executed using a PHP interpreter on the host
+system, the database might not be accessible.
+
+The error usually looks something like this:
+
+```bash
+phpunit
+
+Expected log
+
+exception 'PDOException' with message 'SQLSTATE[HY000] [2002] No such file or directory' in /var/www/magento2/vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php:129
+```
+
+There are many ways this problem can be resolved, but the easiest is to run the tests in the virtual machine as well.
+
+<a href="#running-integration-tests-in-phpstorm"></a>
+## Running Integration Tests in PhpStorm
+
+When writing new integration tests or during debugging, it is convenient to execute tests from within the PhpStorm IDE.
+Ensure you have prepared the integration test environment before starting.
+
+<a href="#creating-an-integration-test-run-configuration"></a>
+### Creating an Integration Test Run Configuration
+
+Setting up a run configuration for integration tests is very similar to creating a run configuration for unit tests. See
+Running Unit Tests in PhpStorm for instructions on creating a basic run configuration. Then, configure the integration
+test to use the configuration file.
+
+<a href="#using-the-integration-test-configuration-file"></a>
+### Using the Integration Test Configuration File
+
+The only difference in the run configuration is that the integration test `phpunit.xml.dist` or `phpunit.xml`
+configuration file from the `dev/tests/integration` directory must be selected.
+
+<a href="#integration-tests-file-structure"></a>
+## Integration Tests File Structure
+
+The root folder for the Magento integration tests suite — `<magento_root>/dev/tests/integration` — contains the
+following sub-folders and files:
+
+- `framework/` – Integration testing framework scripts, configuration files, and classes.
+- `Magento/` – A set of classes that implement the Magento integration tests framework.
+- `bootstrap.php` – The PHPUnit bootstrap script.
+- `etc/install-config-<db_vendor>.php` – A configuration file that provides values for installing the Magento
+  application.
+- `testsuite/` – The test suite.
+- `tmp/` – A writable directory for storing temporary data during test execution.
+- `sandbox-<hash>/` – The folder where each Magento instance stores temporary and configuration data.
+- `phpunit.xml.dist` – A PHPUnit configuration file.
+
+<a href="#using-data-providers-in-integration-tests"></a>
```

Comment:
Although it's a nice feature of PHPUnit, data providers don't belong in this document. I think it leads the reader away from the subject of running tests.